pycaqtimage/pycaqtimage.sip

```diff
@@ -0,0 +1,727 @@
+
+  imageBuffer->isColor      = 1;
+   
+  PyObject* pyfunc = PyCapsule_New(func, PYC_CB, NULL);
+  PyCapsule_SetContext(pyfunc, (void *)imageBuffer);
+  return pyfunc;
+}        
+
+static void _findMinMax(ImageBuffer* imageBuffer)
+{
+    imageBuffer->iProjXmin = 1<<15;
+    imageBuffer->iProjXmax = -1;
+    imageBuffer->iProjYmin = 1<<15;
+    imageBuffer->iProjYmax = -1;
+    
+    for (int iX=0; iX < imageBuffer->imgwidth; ++iX) {
+	int iValue = imageBuffer->projSumX[iX];
+	if (iValue < imageBuffer->iProjXmin)
+	    imageBuffer->iProjXmin = iValue;
+	if (iValue > imageBuffer->iProjXmax)
+	    imageBuffer->iProjXmax = iValue;      
+    }
+    
+    for (int iY=0; iY < imageBuffer->imgheight; ++iY) {
+	int iValue = imageBuffer->projSumY[iY];
+	if (iValue < imageBuffer->iProjYmin)
+	    imageBuffer->iProjYmin = iValue;
+	if (iValue > imageBuffer->iProjYmax)
+	    imageBuffer->iProjYmax = iValue;
+    }        
+    
+    if ( imageBuffer->iProjXmax == -1 ) {
+	imageBuffer->iProjXmin = 0;
+	imageBuffer->iProjXmax = 1;
+    } else {
+	if ( imageBuffer->iProjXmin > 0 )
+	    --imageBuffer->iProjXmin;
+	++imageBuffer->iProjXmax;
+    }
+    
+    if ( imageBuffer->iProjYmax == -1 ) {
+	imageBuffer->iProjYmin = 0;
+	imageBuffer->iProjYmax = 1;
+    } else {
+	if ( imageBuffer->iProjYmin > 0 )
+	    --imageBuffer->iProjYmin;
+	++imageBuffer->iProjYmax;
+    }  
+}
+
+static void _computeRoiProj(ImageBuffer* imageBuffer, QRectF* rectRoi)
+{
+    double*   projSumX  = imageBuffer->projSumX;
+    double*   projSumY  = imageBuffer->projSumY;
+    int       width     = imageBuffer->imgwidth;
+    int       height    = imageBuffer->imgheight;
+
+    for (int iX = 0; iX < width;  projSumX[iX++] = 0);
+    for (int iY = 0; iY < height; projSumY[iY++] = 0);
+
+    /*
+     * Arrange that x1 < x2 and y1 < y2 and they are all in bounds!
+     */
+    int x1  = (int) rectRoi->x();
+    int x2  = x1 + (int) rectRoi->width() - 1;
+    if ( x1 > x2 ) { int xtmp = x1; x1 = x2; x2 = xtmp; }
+    if ( x1 < 0 )
+        x1 = 0;
+    else if ( x1 >= width ) 
+	x1 = width-1;
+    if ( x2 < 0 ) 
+	x2 = 0;
+    else if ( x2 >= width )
+	x2 = width-1;    
+
+    int y1  = (int) rectRoi->y();
+    int y2  = y1 + (int) rectRoi->height() - 1;
+    if ( y1 > y2 ) { int ytmp = y1; y1 = y2; y2 = ytmp; }
+    if ( y1 < 0 )
+	y1 = 0;
+    else if ( y1 >= height ) 
+	y1 = height-1;
+    if ( y2 < 0 ) 
+	y2 = 0;
+    else if ( y2 >= height ) 
+	y2 = height-1;
+  
+    uint32_t* const pImgValue       = imageBuffer->imageData;
+    uint64_t  u64PixelSum           = 0;
+    uint64_t  u64PixelSqSum         = 0;
+    uint32_t* pPixelLineStart       = pImgValue + y1 * width + x1;
+
+    for (int iY = y1; iY <= y2; ++iY, pPixelLineStart += width) {
+	uint32_t* pPixel = pPixelLineStart;
+	for (int iX = x1; iX <= x2; ++iX, ++pPixel) {
+	    int iValue = *pPixel;
+	    if ( iValue >= 0x10000) {
+		printf("Pixel value (%d,%d) too large: value 0x%x\n", iX, iY, iValue);
```

Comment:
For all these printfs, fprintf to stderr at least?

Reply:
Agreed.